### DISCUSSION TITLE
How many pact tests? Not very many unless I'm missing something.

### DISCUSSION THREAD

**Discussion by theGhostJW** [pact-foundation/pact-js#859] [Q&A]
Hi,
I watched the presentation on [writing consumer tests](https://docs.pact.io/consumer) by @bethesque and drawn the conclusion that pact testing is about "structural" testing of contracts and you would need relatively few tests per client to achieve the objectives of pact testing.

The number of tests written would be roughly proportional to the number of distinct structures of request / response pairs the client code would expect. Maybe a full valid request, a minimal valid request and requests where sub-sections of the response are present / missing based on some business rule. Also known classes of error where the error response structure or error code are different from each other. These tests would be further reduced to those requests / response pairs that affect the client implementation. If the client only used 3 fields on a response, there would be no assertions applied against any other fields and no pact needed that doesn't affect the structure of those fields.

If I had a tax service as part of an enterprise service application, testing the tax rate, offsets and other calculations would be the remit of the team looking after the tax service. Unit / property based tests could be executed by developers and component level tests by QAs within the tax service team. 

Pact tests would verify the consumer of this service received the rebate amount, and taxable income fields as expected in the service response (producer test) and the consumer code could actually parse this response (consumer test validation against a mock expected response).

It would be technically possible to execute very specific functional tests, such as bounds tests on income levels, by writing consumer pacts. This would be using pacts as a Trojan horse for tests that should be done by the tax service team and as such is an anti pattern.

Are my ramblings above somewhat in line with the spirit of pact testing or have I misunderstood something?

**Comment by portyu9** [comment]
Hi @theGhostJW 

Yes — your interpretation is very much in line with the intended use of Pact.

The main refinement I would make is that the number of Pact tests isn't necessarily based only on the number of different **JSON structures**. It should be based on the number of materially different **consumer behaviours/contracts**.

A useful question for each potential interaction is:

> If I remove this scenario, what consumer bug or provider compatibility break could now go undetected?

If the answer is "none", you probably don't need that Pact interaction.

For example, if the consumer behaves differently for:

```text
200 → parse and return a domain object
404 → return "not found"
400 → return a validation error
500 → throw/retry
```

then those can legitimately be separate Pact scenarios even if some of their response bodies have similar structures.

But your tax example is exactly the right distinction.

Things such as:

```text
tax calculation correctness
income boundary values
offset calculations
rounding rules
business-rule combinations
```

belong in the **tax provider's own unit/component/functional tests**.

The Pact should instead protect what the consumer actually relies on, for example:

```text
Given a tax calculation is available
When the consumer requests it
Then the response contains:
  taxableIncome: number
  rebateAmount: number
```

If those are the only response fields the consumer reads, those are generally the only fields the consumer contract needs to describe. Adding every field returned by the provider would unnecessarily couple the consumer to fields it does not care about.

Similarly, you would not normally write Pact tests for every possible tax boundary simply to prove that the provider calculates tax correctly. That would be using consumer contracts as provider functional tests.

An extra Pact scenario is justified when the **consumer's expectation or behaviour changes**, for example:

```text
rebate is present
rebate is absent
```

if your consumer actually has different logic for those cases.

So a healthy Pact suite can indeed be surprisingly small compared with the provider's functional test suite.

A good division is:

```text
Consumer Pact tests
- Does my real API client send the request I expect?
- Can it handle the response shapes/statuses it relies on?
- Would a provider change break this consumer?

Provider verification
- Can the provider satisfy those consumer contracts?

Provider unit/component/functional tests
- Is the tax calculation itself correct?
- Are all business rules and boundaries implemented correctly?
- Are the required side effects actually performed?
```

Trying to move that last category into Pact would create lots of interactions without improving contract confidence, and would unnecessarily constrain the provider.

So yes: **few, consumer-relevant interactions are generally preferable to using Pact as a comprehensive functional API test suite.**

Please mark this answer as accepted if it helped, thank you.
